### DISCUSSION TITLE
Puppeteer timeouts when rendering

### DISCUSSION THREAD

**Discussion by ferferga** [lowlighter/metrics#1309] [Q&A] [answered]
[Before, I had 2 steps for metrics generation](https://github.com/ferferga/ferferga/blob/8dab8dbcbe8d75fafc229ce0035e22000e7ed3af/.github/workflows/metrics.yml) (one for the left and one for the right). I'm trying to merge them in a single step [as you can see in current master](https://github.com/ferferga/ferferga) *(if you're an user reading this from the future, be aware I usually force push this repository, so what's currently in master will change)*. I tried disabling all plugins and enabling them back one by one manually and it breaks as soon as I enable the achievements one. Commenting the ``plugin_achievements_*`` config completely fixes the issue

Here's an screenshot of whats the last thing logged (GitHub seems to clear some logs from the workflow run): 
![image](https://user-images.githubusercontent.com/10274099/206459671-cb10259e-f4eb-40fd-8fae-d4c4484733ee.png)

Here is one of the failing workflow runs:
https://github.com/ferferga/ferferga/actions/runs/3648218293

**Comment by lowlighter** [comment, marked as answer]
I tried to replicate your issue and couldn't get the same unfortunately 😕 

Even though I was also getting the negative rect it still rendered (I was testing in svg output though) 
```
  metrics/svg/resize > loading svg
  metrics/svg/resize > puppeteer > Error: <rect> attribute width: A negative value is not valid. ("-2.303664921465969")
  metrics/svg/resize > puppeteer > Error: <rect> attribute width: A negative value is not valid. ("-2.8002213613724405")
  metrics/svg/resize > loaded svg successfully
  metrics/svg/resize > resizing svg
  metrics/svg/resize > puppeteer > animations are enabled
  metrics/svg/resize > puppeteer > bounds width=980, height=1045.765625
  metrics/svg/resize > puppeteer > bounds after applying padding width=980 (*1+0), height=1046 (*1+0)
  metrics/svg/resize > rendering complete
  metrics/compute/ferferga > success
```

I fixed the negative values and tried again in both svg and png this time and the output was succesful so I'm not sure what's causing this.

Could you try using the svg output and see if you still encouter your error ? 
The achievement plugin has some CSS animations so maybe it's making the png rendering fail in some case 🤔 ?

**Reply by ferferga** [reply]
Still the same thing unfortunately: https://github.com/ferferga/ferferga/actions/runs/3673794083/jobs/6211253639

By any chance, did you try with the same config I used (but with your PAT of course)?

**Reply by lowlighter** [reply]
Yeah I copied the one you had on your repository and the result looks like this (image scaled down):
<img src="https://user-images.githubusercontent.com/22963968/207211957-30279f6c-b396-4513-8997-a993a9f99521.png" width="600">

___

Since I'm not able to replicate your issue I'm not sure what's causing it unfortunately 😕 

I can only guess but from the logs you have, it seems to be blocked around the final rendering part:
https://github.com/lowlighter/metrics/blob/f99da6c94154d2a35384b004b2cfbd7cc83830b9/source/app/metrics/utils.mjs#L533-L541

Maybe one of the events from `["load", "domcontentloaded", "networkidle2"]` doesn't trigger properly which is why it's making puppeteer timeout 🤔 ?

**Reply by ferferga** [reply]
I'm clueless but, before I start a fork to test code changes, I think if we had more information about what happens with Puppeeteer we would have a great starting point. Would you mind making a commit that [passes the debug environment variable to it](https://pptr.dev/guides/debugging/#log-devtools-protocol-traffic) when metrics is also running in debug mode?

**Reply by lowlighter** [reply]
With #1317 you'll have access to new [`debug_flags`](https://github.com/lowlighter/metrics/blob/master/source/plugins/core/README.md#debug_flags) to debug puppeteer 👍 

- Passing `--puppeteer-debug` will enable the debug logging by setting the env var you linked
- Passing one or more of `--puppeteer-wait-load`, `--puppeteer-wait-domcontentloaded`, `--puppeteer-wait-networkidle0`, `--puppeteer-wait-networkidle2` will override the default events puppeteer is waiting for rendering

Hopefully it can help to solve your issue 🤞

**Reply by ferferga** [reply]
So it looks like the issue resolved itself, even without debug flags enabled. I went commit by commit to find out more hints:

* **master**: https://github.com/ferferga/ferferga/actions/runs/3693338901 🟢
* **master *(with --pupeeteer-debug enabled)***: https://github.com/ferferga/ferferga/actions/runs/3693455916 🟢
* **latest**: https://github.com/ferferga/ferferga/actions/runs/3693960368 ❌

*At this point, beside the commit tag, all the tests were done also with the ``use_prebuilt_image: no`` flag*

* 77fd4d044d73be31d346b22b88a384d7a0c764a1: https://github.com/ferferga/ferferga/actions/runs/3693547530 🟢
* 73c8e87dc7bb8f874eef1585add26cfef81fbd0a: https://github.com/ferferga/ferferga/actions/runs/3693645492 🟢
* 3a4559e6ed4821c9659a92f696e6fbc920309709: https://github.com/ferferga/ferferga/actions/runs/3693741084 🟢
* a90044266c8c0e803008b8c7b4fae353d35368ed: https://github.com/ferferga/ferferga/actions/runs/3693814533 🟢
* **latest**: https://github.com/ferferga/ferferga/actions/runs/3693888049 🟢

The important thing is that now I have fully working metrics (with ``use_prebuilt_image: no``, but better that than nothing) so thank you very much for your help 🙌.
However, I don't know how to further diagnose which commit might have fixed this, as all of them seem to pass the checks perfectly fine. My guess is that the problem exists in a dependency inside the Docker container (and that's why latest without prebuilt works but latest with prebuilt **doesn't**).

**Reply by lowlighter** [reply]
Yeah I think you're probably right, the latest release probably have a faulty dependency packaged in which could explain this.
Thanks a lot for looking into it, glad you solved the issue 😄

**Reply by ferferga** [reply]
Any chance you can rebuild and retag the image or shall I wait for the next update if I want to stick with **latest**?

**Reply by lowlighter** [reply]
I prefer releasing a new version rather than altering an already released version (even if it's broken)

#1315 is pretty high on my todo-list (planning to work on it this week-end) and I plan to make a release once it's patched so `@latest` should be updated in the next few days 👍

**Reply by ferferga** [reply]
@lowlighter The new release now includes the fixed dependency: https://github.com/ferferga/ferferga/actions/runs/3755998187/jobs/6381582666

Thank you very much for all your support throughout this issue! Please, mark my message as answer, GH doesn't allow me to do that.
